commons/src/main/java/org/infinispan/commons/api/functional/Traversable.java

```diff
@@ -0,0 +1,215 @@
+package org.infinispan.commons.api.functional;
+
+import org.infinispan.commons.api.functional.Closeables.CloseableSpliterator;
+import org.infinispan.commons.util.CloseableIterator;
+
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.BinaryOperator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * Unsorted traversable stream. Design is inspired on {@link java.util.stream.Stream}.
+ * Some functions that would be typically found in {@link java.util.stream.Stream}
+ * are not included here:
+ *
+ * <ul>
+ *    <li>{@link java.util.stream.Stream#distinct()}: This operation would be
+ *    extremely costly in a distributed environment as all data up to that
+ *    point of the traversable stream must be read by a single node.
+ *    Due to the memory and performance constraints this method wouldn't be supported.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#sorted()}: Providing a sorted
+ *    traversable stream in a distributed environment is an expensive operation.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#limit(long)}: In a distributed
+ *    environment this would require sequential requests for each node to
+ *    verify that we haven't reached the limit yet. This could be optimized
+ *    depending on the terminator. This operation makes most sense when
+ *    combined with sorted streams.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#sorted()}: Operation that makes
+ *    most sense when combined with sorting.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#forEachOrdered(Consumer)}:
+ *    Sorting related.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#toArray()}: Creating an array
+ *    requires size to be computed first, and then populate it, making it a
+ *    potentially expensive operation in a distributed operation. Also,
+ *    there are other ways to achieve this, such as mapping to a variable
+ *    length collection and then narrowing it down to an array?
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#collect(java.util.stream.Collector)}:
+ *    This method is designed to work with {@link java.util.stream.Collectors},
+ *    whose static methods are used as helpers functions, but these are not
+ *    Serializable, and we'd need to investigate whether we can provide
+ *    marshallable versions. Also, {@link java.util.stream.Collectors} brings
+ *    in a lot of baggage. Instead, {@link #collect(Supplier, BiConsumer, BiConsumer)}
+ *    should be used.
+ *    </li>
+ *    <li>{@link java.util.stream.Stream#findFirst()}: The only difference
+ *    with {@link #findAny()} is ordering, so since we do not want to enter
+ *    the ordering business, we leave it out.</li>
```

Comment:
@danberindei My thinking here that we don't need to support everything that's in stream, particularly those operations that can be expensive or can potentially blow up things. Why not? Although they might have their uses, say of sorting, we're still providing a ConcurrentMap implementation and by extension we have to support Stream operations. So, if anyone wants to do sorting, they can get a ConcurrentMap view of our cache, which is what Infinispan Cache exposes, and then use that. In Traversable, I've tried to provide the essential functionality with the knowledge that if anyone's missing anything that they find in Stream, they can still get that. With that in mind, given sorting's caveats, I'd still leave it out. We can always add it in later on.